### ISSUE TITLE
`beforeEach` and `afterEach` are running in nested suites/contexts

### ISSUE THREAD

**Issue by twolfson** [mochajs/mocha#911]
There is no documentation on _when_ a `beforeEach` or `afterEach` will run. My intuition states that it should be run `before`/`after` every `describe`/`it` block in the current context completes.

However, the behavior I am noticing, is that `beforeEach` and `afterEach` are run `before`/`after` _every_ `it` block in the current context and all nested contexts.

I have created a proof of concept to demonstrate that the noticed behavior is occurring: https://gist.github.com/twolfson/5883057#file-test-js

For convenience, I will copy/paste the script and output here:
### Script

```
describe('An afterEach hook', function () {
  afterEach(function () {
    console.log('afterEach run!');
  });

  before(function () {
    console.log('before run!');
  });

  describe('in some nested contexts', function () {
    before(function () {
      console.log('nested before run!');
    });

    it('runs after this block', function () {
      console.log('nested it run!');

    });

    it('runs again after this block', function () {
      console.log('second nested it run!');
    });
  });
});
```
### Output

```
before run!
nested before run!
nested it run!
afterEach run!
second nested it run!
afterEach run!
```

The behavior I intuitively anticipate is `afterEach` to run once, after the `second nested it`.

```
before run!
nested before run!
nested it run!
second nested it run!
afterEach run!
```

My use case for the intuitive functionality is cleaning out the `this` context after a context completes. The current implementation will clean out the context after every assertion.

``` js
afterEach(function () {
  var key;
  for (key in this) {
    delete this[key];
  }
});

describe('A banana', function () {
  before(function () {
    this.banana = new Banana();
  });

  describe('when peeled', function () {
    before(function () {
      this.peeledBanana = this.banana.peel();
    });

    it('is white', function () {
      assert.strictEqual(this.peeledBanana.color, 'white');
      // `afterEach` is invoked here, cleaning out `this`
    });

    it('is soft', function () {
      // `this.peeledBanana` is no longer defined since `afterEach` cleaned it out
      assert.strictEqual(this.peeledBanana.hardness, 'soft');
    });
  });
});
```

A comparable example to this would be opening a database transaction in each new context of a test suite.

I can work around it by invoking `after` in each nested context but I feel like there could be a hook in the framework.

**Comment by RathaKM** [comment]
Is there a way to run hook methods in the describe() level instead of it() level? It might be useful in some scenarios.

**Comment by boneskull** [comment]
@RathaKM What's the use case?

**Comment by cpoonolly** [comment]
Would the following be a good example of the usefulness of this? @RathaKM feel free to correct.
```
describe('A banana', function () {
  var banana;

  beforeEach(function () {
    banana = new Banana();
  });

  describe('when peeled', function () {
    var peeledBanana;

    before(function () {
      // Lets assuming peel() HAS side-effects and doesn't return a new object.
      peeledBanana = banana.peel();
    });

    it('is white', function () {
      assert.strictEqual(peeledBanana.color, 'white');
    });

    it('is soft', function () {
      assert.strictEqual(peeledBanana.hardness, 'soft');
    });
  });

  describe('when set on FIRE', function () {
    var flamingBanana;

    before(function () {
      // Same assumption as above
      flamingBanana = banana.setOnFire();
    });

    it('is hot', function () {
      assert.isAbove(flamingBanana.temperature, 9000);
    });
  });
});
```

**Comment by RathaKM** [comment]
@cpoonolly: thanks for your comments and bring it back. :) Your example is what logically expected to work in the describe level(the BeforeEach). But, this could be achieved with 'before()' within each describe.

**Comment by cloutiertyler** [comment]
Has there been any additional discussion on this?

**Comment by marqu3z** [comment]
I really don't understand why is not possible to have something like this:

```javascript
describe('Create Article', () => {
  beforeEach(() => console.log('CLEAN DB'))

  context('When Article is valid', () => {
    before(() => console.log("INSERTING VALID USER"))

    it('should not fail')
    it('should have some properties')    
    it('should send an email')  
  })

  context('When Article is not valid', () => {
    before(() => console.log("INSERTING VALID USER"))

    it('should fail')
    it('should not send an email')  
  })
})
```

 - Clean the DB **before each** `context` block.
 - Set up the scenario **before all** `it` blocks

Isn't this a legit use case?

**Comment by ORESoftware** [comment]
@marqu3z why not just clean the DB in the `before` hook of the context block? that should work right?

**Comment by ORESoftware** [comment]
@RathaKM 

are you registering your it() test cases asynchronously? If so, that is not allowed. You must register `describe/after/afterEach/before/beforeEach` all synchronously.

```js
// event loop tick X
 dataDriven(testData, function () {
       // it() must be called in the same event loop tick as X above
            it('expecting updated testData for UK', function (ctx, done) {
                assert.equal(ctx.country, 'UK');
                done();
            });
        });
```

**Comment by makrandgupta** [comment]
I'm in the exact same scenario @marqu3z described. Would love to see support for a `describe` level `beforeEach`

**Comment by marqu3z** [comment]
To everyone with this common scenario i created this helper function that iterates over all suites in the current block and add a `beforeAll` hook to each of them:

```javascript
function beforeEachSuite (fn) {
  before(function () {
    let suites = this.test.parent.suites || []
    suites.forEach(s => {
      s.beforeAll(fn)
      let hook = s._beforeAll.pop()
      s._beforeAll.unshift(hook)
    })
  })
}
```

This unlock the use case i describe almost an year ago:

```javascript
describe('Create Article', () => {
  beforeEachSuite(() => console.log('CLEAN DB'))

  context('When Article is valid', () => {
    before(() => console.log("INSERTING VALID ARTICLE"))

    it('should not fail')
    it('should have some properties')    
    it('should send an email')  
  })

  context('When Article is not valid', () => {
    before(() => console.log("INSERTING NOT VALID ARTICLE"))

    it('should fail')
    it('should not send an email')  
  })
})
```
Still don't get why this use case is not considered useful, or legit. 
Maybe this solution can be turned in a PR for a `beforeEachSuite` and `afterEachSuite`

**Comment by elitan** [comment]
Yes please we need this.

There are many use cases like https://github.com/mochajs/mocha/issues/911#issuecomment-316736991

**Comment by mehtamanan** [comment]
+1

**Comment by sinkovsky** [comment]
+1

**Comment by owenfarrell** [comment]
I just stumbled on this thread looking for something similar. So to anyone else who stumbles on this post...

I've adapted the approach that @marqu3z outlined in to its own NPM package: [`mocha-suite-hooks`](https://www.npmjs.com/package/mocha-suite-hooks).

So if you find yourself in a situation where `before` isn't enough, and `beforeEach` is too much, give `beforeSuite` a try.

The package is still pretty raw, so any feedback would be greatly appreciated!
